exonum-java-binding/core/rust/src/proxy/runtime.rs

```diff
@@ -0,0 +1,561 @@
+                    &[
+                        JValue::from(view_handle),
+                    ],
+                ),
+            );
+            Ok(())
+        }));
+    }
+
+    fn after_commit(
+        &self,
+        _dispatcher: &DispatcherSender,
+        snapshot: &Snapshot,
+        service_keypair: &(PublicKey, SecretKey),
+        _tx_sender: &ApiSender,
+    ) {
+        unwrap_jni(self.exec.with_attached(|env| {
+            let view_handle = to_handle(View::from_ref_snapshot(snapshot));
+            let validator_id = Self::validator_id_or(snapshot, &service_keypair.0, -1);
+            let height:u64 = CoreSchema::new(snapshot).height().into();
+
+            panic_on_exception(
+                env,
+                env.call_method_unchecked(
+                    self.runtime_adapter.as_obj(),
+                    runtime_adapter::after_commit_id(),
+                    JavaType::Primitive(Primitive::Void),
+                    &[
+                        JValue::from(view_handle),
+                        JValue::from(validator_id),
+                        JValue::from(height as i64),
+                    ],
+                ),
+            );
+            Ok(())
+        }));
+    }
+
+    fn notify_api_changes(&self, context: &ApiContext, changes: &[ApiChange]) {
+        let added_instances_ids: Vec<i32> = changes.iter()
+            .filter_map(|change| {
+                if let ApiChange::InstanceAdded(instance_id) = change {
+                    Some(*instance_id as i32)
+                } else {
+                    None
+                }
+            })
+            .collect();
+
+        if added_instances_ids.is_empty() {
+            return;
+        }
+
+        let node = NodeContext::new(self.exec.clone(), context.clone());
+
+        unwrap_jni(self.exec.with_attached(|env| {
+            let node_handle = to_handle(node);
+            let ids_array = env.new_int_array(added_instances_ids.len() as i32)?;
+            env.set_int_array_region(ids_array, 0, &added_instances_ids)?;
+            let service_ids = JObject::from(ids_array);
+
+            panic_on_exception(
+                env,
+                env.call_method_unchecked(
+                    self.runtime_adapter.as_obj(),
+                    runtime_adapter::connect_apis_id(),
+                    JavaType::Primitive(Primitive::Void),
+                    &[
+                        JValue::from(service_ids),
+                        JValue::from(node_handle),
+                    ],
+                ),
+            );
+            Ok(())
+        }));
+    }
+}
+
+impl fmt::Debug for JavaRuntimeProxy {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "JavaRuntimeProxy()")
+    }
+}
+
+/// Artifact identification properties within `JavaRuntimeProxy`
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct JavaArtifactId(String);
+
+impl ToString for JavaArtifactId {
```

Comment:
According to docs from the standard library this trait should not be implemented directly:

>  This trait is automatically implemented for any type which implements the
[`Display`] trait. As such, `ToString` shouldn't be implemented directly:
[`Display`] should be implemented instead, and you get the `ToString`
implementation for free.